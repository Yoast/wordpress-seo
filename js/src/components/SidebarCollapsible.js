import React from "react";
import { Collapsible } from "@yoast/components";

/**
 * Sidebar Collapsible component with default padding and separator
 *
 * @param {Object} props The properties for the component.
 *
 * @returns {ReactElement} The Collapsible component.
 */
const SidebarCollapsible = ( props ) => {
	return <Collapsible hasPadding={ true } hasSeparator={ true } { ...props } />;
};

export default SidebarCollapsible;
