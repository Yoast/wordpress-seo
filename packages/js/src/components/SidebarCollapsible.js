import { useState } from "@wordpress/element";
import { SvgIcon } from "@yoast/components";
import PropTypes from "prop-types";

/**
 * Sidebar Collapsible component with default padding and separator
 *
 * @param {Object} props The properties for the component.
 *
 * @returns {wp.Element} The Collapsible component.
 */
const SidebarCollapsible = ( props ) => {
	const [ isOpen, toggleOpen ] = useState( false );

	const {
		prefixIcon,
	} = props;

	/**
	 * Toggles the SidebarCollapsible open and closed state.
	 *
	 * @returns {void}
	 */
	function handleClick() {
		toggleOpen( ! isOpen );
	}

	return <div className={ `yoast components-panel__body ${ isOpen ? "is-opened" : "" }` }>
		<h2 className="components-panel__body-title">
			<button
				onClick={ handleClick }
				className="components-button components-panel__body-toggle"
			>
				<span
					className="yoast-icon-span"
					style={ { fill: `${ prefixIcon && prefixIcon.color || "" }` } }
				>
					{
						prefixIcon && <SvgIcon
							icon={ prefixIcon.icon }
							color={ prefixIcon.color }
							size={ prefixIcon.size }
						/>
					}
				</span>
				<span className="yoast-title-container">
					<div className="yoast-title">{ props.title }</div>
					<div className="yoast-subtitle">{ props.subTitle }</div>
				</span>
				<span className="yoast-chevron" aria-hidden="true" />
			</button>
		</h2>
		{ isOpen && props.children }
	</div>;
};

export default SidebarCollapsible;

SidebarCollapsible.propTypes = {
	title: PropTypes.string.isRequired,
	children: PropTypes.oneOfType( [
		PropTypes.node,
		PropTypes.arrayOf( PropTypes.node ),
	] ).isRequired,
	prefixIcon: PropTypes.object,
	subTitle: PropTypes.string,
};

SidebarCollapsible.defaultProps = {
	prefixIcon: null,
	subTitle: "",
};
