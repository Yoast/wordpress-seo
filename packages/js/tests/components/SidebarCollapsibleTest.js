import SidebarCollapsible from "../../src/components/SidebarCollapsible";
import { fireEvent, render, screen } from "../test-utils";

const defaultArgs = {
	title: "Test title",
	children: <div id="inner-div">Test children</div>,
};

describe( "SidebarCollapsible", () => {
	it( "is opened", async() => {
		const { container } = render( <SidebarCollapsible { ...defaultArgs } /> );
		const button = screen.getByRole( "button" );
		fireEvent.click( button );
		const innerDiv = container.querySelector( "#inner-div" );
		expect( innerDiv ).toBeInTheDocument();
		expect( container ).toMatchSnapshot();
	} );

	it( "is closed", async() => {
		const { container } = render( <SidebarCollapsible { ...defaultArgs } /> );
		const innerDiv = container.querySelector( "#inner-div" );
		expect( innerDiv ).not.toBeInTheDocument();
		expect( container ).toMatchSnapshot();
	} );

	it( "has beta badge", async() => {
		const { container } = render( <SidebarCollapsible hasBetaBadgeLabel={ true } { ...defaultArgs } /> );
		const badge = container.querySelector( ".yoast-beta-badge" );
		expect( badge ).toBeInTheDocument();
	} );

	it( "has svg prefix icon", async() => {
		const { container } = render( <SidebarCollapsible prefixIcon={ { icon: "gear", color: "green" } } { ...defaultArgs } /> );
		const icon = container.querySelector( "svg" );

		expect( icon ).toBeInTheDocument();
	} );

	it( "has id associated to inner button", async() => {
		const { container } = render( <SidebarCollapsible buttonId="test-button-id" { ...defaultArgs } /> );
		const button = container.querySelector( "#test-button-id" );
		expect( button ).toBeInTheDocument();
	} );
} );
