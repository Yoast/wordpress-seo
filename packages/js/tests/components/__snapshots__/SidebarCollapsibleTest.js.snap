// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarCollapsible is closed 1`] = `
<div>
  <div
    class="yoast components-panel__body "
  >
    <h2
      class="components-panel__body-title"
    >
      <button
        class="components-button components-panel__body-toggle"
        type="button"
      >
        <span
          class="yoast-icon-span"
        />
        <span
          class="yoast-title-container"
        >
          <div
            class="yoast-title"
          >
            Test title
          </div>
          <div
            class="yoast-subtitle"
          />
        </span>
        <span
          aria-hidden="true"
          class="yoast-chevron"
        />
      </button>
    </h2>
  </div>
</div>
`;

exports[`SidebarCollapsible is opened 1`] = `
<div>
  <div
    class="yoast components-panel__body is-opened"
  >
    <h2
      class="components-panel__body-title"
    >
      <button
        class="components-button components-panel__body-toggle"
        type="button"
      >
        <span
          class="yoast-icon-span"
        />
        <span
          class="yoast-title-container"
        >
          <div
            class="yoast-title"
          >
            Test title
          </div>
          <div
            class="yoast-subtitle"
          />
        </span>
        <span
          aria-hidden="true"
          class="yoast-chevron"
        />
      </button>
    </h2>
    <div
      id="inner-div"
    >
      Test children
    </div>
  </div>
</div>
`;
